### PR TITLE
Adding compatibility mode for unpacker

### PR DESCRIPTION
**Pull request by umverteiler**
Adding compatibility mode for unpacker (type raw contains strings or binary data).

### CODE REVIEW

**Comment by a2** [comment]
I know this is a small thing but can you move the `else {` lines onto the `}` lines (`} else {`)?

**Comment by umverteiler** [comment]
No problem, done.

File: MessagePack/Source/MessagePack.swift [hunk 1]
```diff
@@ -616,6 +630,12 @@ extension MessagePackValue {
     /// The contained string if `.String`, `nil` otherwise.
     public var stringValue: Swift.String? {
         switch self {
+        case .Binary(let data):
+            if let string = NSString(data: data, encoding:NSASCIIStringEncoding) {
```
Comment (a2) [inline MessagePack/Source/MessagePack.swift]: Should we default to `NSUTF8StringEncoding`?

Comment (umverteiler) [inline MessagePack/Source/MessagePack.swift]: Yes, you're right - well actually the user should decide for his / her environment, but UTF-8 would be the best default setting.

File: MessagePack/Source/MessagePack.swift [hunk 2]
```diff
@@ -616,6 +630,14 @@ extension MessagePackValue {
     /// The contained string if `.String`, `nil` otherwise.
     public var stringValue: Swift.String? {
         switch self {
+        case .Binary(let data):
+            var string = ""
+            let ptr = UnsafePointer<UInt8>(data.bytes)
+            let bytes = UnsafeBufferPointer<UInt8>(start:ptr, count:data.length)
```
Comment (a2) [inline MessagePack/Source/MessagePack.swift]:
``` swift
let bytes = UnsafeBufferPointer<UInt8>(start: ptr, count: data.length)
```

File: MessagePack/Source/MessagePack.swift [hunk 3]
```diff
@@ -616,6 +630,14 @@ extension MessagePackValue {
...
+            for var i = 0; i < data.length; i++ {
```
Comment (a2) [inline MessagePack/Source/MessagePack.swift]:
``` swift
for byte in bytes { ... }
```

File: MessagePack/Source/Helpers.swift [hunk 4]
```diff
@@ -87,12 +87,12 @@ func joinData<G: GeneratorType where G.Element == UInt8>(inout generator: G, len
 
     :returns: An `Array`, or `nil` if the generator runs out of elements.
 */
-func joinArray<G: GeneratorType where G.Element == UInt8>(inout generator: G, length: Int) -> [MessagePackValue]? {
+func joinArray<G: GeneratorType where G.Element == UInt8>(inout generator: G, length: Int, compatibility: Bool) -> [MessagePackValue]? {
```
Comment (a2) [inline MessagePack/Source/Helpers.swift]: Retain the argument name requirement:

``` swift
#compatibility: Bool
```

File: MessagePack/Source/Helpers.swift [hunk 5]
```diff
@@ -110,9 +110,9 @@ func joinArray<G: GeneratorType where G.Element == UInt8>(inout generator: G, le
 
     :returns: A `Dictionary`, or `nil` if the generator runs out of elements.
 */
-func joinMap<G: GeneratorType where G.Element == UInt8>(inout generator: G, length: Int) -> [MessagePackValue : MessagePackValue]? {
+func joinMap<G: GeneratorType where G.Element == UInt8>(inout generator: G, length: Int, compatibility: Bool) -> [MessagePackValue : MessagePackValue]? {
```
Comment (a2) [inline MessagePack/Source/Helpers.swift]: As well as here, and the other functions as well

**Comment by a2** [comment]
This looks fine on the whole. Just a couple of nits :smile_cat:

**Comment by umverteiler** [comment]
OK, changes done.

**Comment by a2** [comment]
Looks great. :smiley:

**Comment by a2** [comment]
Would you mind making similar changes to the `swift-2.0` branch?

**Comment by umverteiler** [comment]
Sure - but first I need to migrate our own Project to Swift 2.0 (should happen in September).

**Pull request merged**